[iOS] Add SPI to vend the find overlay layer

pxlcoder opened the pull request:
#### 90483815379d97337a85f9424fad5d4e64ceb1ae
<pre>
[iOS] Add SPI to vend the find overlay layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=241357">https://bugs.webkit.org/show_bug.cgi?id=241357</a>
rdar://90466280

Reviewed by Simon Fraser, Tim Horton and Wenson Hsieh.

Currently, when searching a conversation in Mail, find overlays for multiple
messages do not animate in at the same time. This issue occurs due to the fact
that each message is in its own web view, backed by its own web process, and
showing the find overlay involves communication across UI and web processes.

In order to ensure the overlays are animated in/out at the same time, give
Mail access to the find overlay layer. The overlays can then be displayed
alongside other Mail content. The overlay layer is exposed through WKWebView
SPI, and _WKFindDelegate methods to inform the client when the overlay is
added/removed.

Alternatives considered:

1. Mail owns the overlay layer and paints highlights themselves. This approach
was not chosen, as WebKit would still need to provide SPI to vend the rects, and
the complexity of implementation increases as the highlight rects could change
position relative to the web view (for example, in subscrollable regions).
In general, this approach would involve significant work in both Mail and WebKit.

2. Mail owns the overlay layer and gives it to WebKit. This approach was not
chosen, as it is incompatible with the existing remote layer tree.

* Source/WebCore/page/PageOverlay.cpp:
(WebCore::PageOverlay::create):
(WebCore::PageOverlay::PageOverlay):
* Source/WebCore/page/PageOverlay.h:
* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::installPageOverlay):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::supportsLayerType):
* Source/WebCore/platform/graphics/GraphicsLayer.h:

Introduce a new layer type that ensures the layer is always tile-backed. This
is necessary for client-added overlays, since the layer type could otherwise
change, leaving the client with the incorrect layer.

(WebCore::GraphicsLayer::setOpacity):

Export this method, so that the find overlay can begin at zero opacity. The
client is responsible for changing the opacity.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayer::supportsLayerType):
(WebCore::GraphicsLayerCA::initialize):
(WebCore::GraphicsLayerCA::adjustCoverageRect const):
(WebCore::GraphicsLayerCA::requiresTiledLayer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::isTiledBackingLayer const):
* Source/WebKit/UIProcess/API/APIFindClient.h:
(API::FindClient::didAddLayerForFindOverlay):
(API::FindClient::didRemoveLayerForFindOverlay):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFindDelegate.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _processWillSwapOrDidExit]):
(-[WKWebView _didCommitLayerTree:]):

Inform the client that the find overlay is available, once it has been created
and parented. Note that the layer properties arrive in a separate transaction,
since the overlay container is created lazily. Consequently, we wait until the
layer is parented. This ensures that the opacity of the layer is actually zero
once the client is notified.

(-[WKWebView _addLayerForFindOverlay]):

Mark the find overlay as pending once we have a layer ID. At this point, the
layer is not yet in the tree.

(-[WKWebView _removeLayerForFindOverlay]):

Do not send a message to the web process if there is no overlay.

(-[WKWebView _layerForFindOverlay]):

Return the find overlay layer using the `_committedFindLayerID`.

* Source/WebKit/UIProcess/Cocoa/FindClient.h:
* Source/WebKit/UIProcess/Cocoa/FindClient.mm:
(WebKit::FindClient::setDelegate):
(WebKit::FindClient::didAddLayerForFindOverlay):
(WebKit::FindClient::didRemoveLayerForFindOverlay):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addLayerForFindOverlay):
(WebKit::WebPageProxy::removeLayerForFindOverlay):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::addLayerForFindOverlay):

Ensure the overlay is always tile-backed. Without this change, the layer ID can
change depending on the size of the web view, and could leave the client with
the incorrect layer.

(WebKit::WebFoundTextRangeController::removeLayerForFindOverlay):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addFindOverlayLayer):
(WebKit::WebPage::removeFindOverlayLayer):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(-[TestFindDelegate setDidAddLayerForFindOverlayHandler:]):
(-[TestFindDelegate didAddLayerForFindOverlayHandler]):
(-[TestFindDelegate setDidRemoveLayerForFindOverlayHandler:]):
(-[TestFindDelegate didRemoveLayerForFindOverlayHandler]):
(-[TestFindDelegate _webView:didAddLayerForFindOverlay:]):
(-[TestFindDelegate _webViewDidRemoveLayerForFindOverlay:]):
(TEST):

Added a test to exercise the new SPI.

Canonical link: <a href="https://commits.webkit.org/252181@main">https://commits.webkit.org/252181@main</a>
</pre>

whsieh commented on Source/WebKit/UIProcess/Cocoa/FindClient.h:
```diff
@@ -57,6 +60,8 @@ class FindClient final : public API::FindClient {
         bool webviewDidCountStringMatches : 1;
         bool webviewDidFindString : 1;
         bool webviewDidFailToFindString : 1;
+        bool webviewDidAddFindOverlayLayer;
```
Nit - should these also be 1-bit-wide, to match the other flags above?

whsieh commented on Source/WebKit/UIProcess/Cocoa/FindClient.mm:
```diff
@@ -67,4 +69,16 @@
+void FindClient::didAddFindOverlayLayer(WebKit::WebPageProxy*, CALayer *layer)
+{
+    if (m_delegateMethods.webviewDidAddFindOverlayLayer)
+        [m_delegate.get() _webView:m_webView didAddFindOverlayLayer:layer];
```
Nit - you can omit the `.get()` here.

whsieh commented on Source/WebKit/UIProcess/Cocoa/FindClient.mm:
```diff
@@ -67,4 +69,16 @@
+void FindClient::didRemoveFindOverlayLayer(WebKit::WebPageProxy*)
+{
+    if (m_delegateMethods.webviewDidRemoveFindOverlayLayer)
+        [m_delegate.get() _webViewDidRemoveFindOverlayLayer:m_webView];
```
(Ditto)

hortont424 commented on Source/WebKit/UIProcess/API/APIFindClient.h:
```diff
@@ -28,6 +28,8 @@
 
 #include <wtf/text/WTFString.h>
 
+OBJC_CLASS CALayer;
```
Should this file contain platform details? Unconditionally? Maybe these should be PlatformLayer?

pxlcoder replied:
Missed this feedback in the first iteration. Updated to use `PlatformLayer` since this file is technically platform-agnostic.

hortont424 commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -193,6 +193,26 @@ void WebFoundTextRangeController::didEndTextSearchOperation()
+        m_findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
+        m_webPage->corePage()->pageOverlayController().installPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::DoNotFade);
+        m_findPageOverlay->layer().setOpacity(0);
+        m_findPageOverlay->layer().setAllowsTiling(false);
```
What's this about! I would expect the find overlay to usually be tiled.

hortont424 replied:
(I see your comment in the commit message, but this seems like a mistake for large screen sizes?)

pxlcoder replied:
Any thoughts on doing the opposite, and always making the layer tiled?

Right now, the layer can be changed depending on its width/height (`GraphicsLayerCA::requiresTiledLayer` via `commitLayerChangesBeforeSublayers`).

hortont424 replied:
Always-tiled is fine, we do that sometimes.

smfr commented on Source/WebCore/page/PageOverlay.h:
```diff
@@ -67,7 +67,12 @@ class PageOverlay final : public RefCounted<PageOverlay> {
     };
 
-    WEBCORE_EXPORT static Ref<PageOverlay> create(Client&, OverlayType = OverlayType::View);
+    enum class RequiresTiling : bool {
```
This name doesn't make it clear if the intent is "always tile", or "avoid tiling".

pxlcoder replied:
Renamed to `AlwaysTileOverlayLayer` for clarity.

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -253,7 +253,8 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
         ScrolledContents,
-        Shape
+        Shape,
+        TiledBacking
```
Tiling is an implementation detail of GraphicsLayer, so I'd prefer not to expose this. I think it would be better for GraphicsLayer to have `setAlwaysUseTiling()` or something.

pxlcoder replied:
We do already have `PageTiledBacking`. Any thoughts on keeping this as-is, or renaming to indicate that this should be used with caution?

`setAlwaysUseTiling()` is not possible with the current architecture, since the LayerType is set at construction time, Can we consider a larger refactor as a follow-up?

smfr replied:
Seems reasonable to just go with TiledBacking. Please move it up below PageTiledBacking.

pxlcoder replied:
Done – thank you!

smfr commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -372,6 +372,9 @@ - (void)_initializeWithConfiguration:(WKWebViewConfiguration *)configuration
     _allowsLinkPreview = linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::LinkPreviewEnabledByDefault);
     _findInteractionEnabled = NO;
 
+    _pendingFindLayerID = 0;
```
Do we also need to zero these out on a web process crash?

pxlcoder replied:
Good point, I did this in `-[WKWebView _processWillSwapOrDidExit]`.

smfr commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -509,7 +509,12 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
 @property (nonatomic, readonly) _UIFindInteraction *_findInteraction WK_API_AVAILABLE(ios(WK_IOS_TBA));
 @property (nonatomic, readwrite, setter=_setFindInteractionEnabled:) BOOL _findInteractionEnabled WK_API_AVAILABLE(ios(WK_IOS_TBA));
 
+@property (nonatomic, readonly) CALayer *_findOverlayLayer WK_API_AVAILABLE(ios(WK_IOS_TBA));
```
Ambiguous name: "find the overlay layer" or "layer for find overlay". _layerForFindOverlay: would avoid the ambiguity.

pxlcoder replied:
Renamed.

smfr commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -509,7 +509,12 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
+
 - (void)_requestRectForFoundTextRange:(UITextRange *)ranges completionHandler:(void (^)(CGRect))completionHandler WK_API_AVAILABLE(ios(WK_IOS_TBA));
+
+- (void)_addFindOverlayLayer WK_API_AVAILABLE(ios(WK_IOS_TBA));
```
addLayerFor/removeLayerForFindOverlay.

pxlcoder committed:
[iOS] Add SPI to vend the find overlay layer
https://bugs.webkit.org/show_bug.cgi?id=241357
rdar://90466280

Reviewed by Simon Fraser, Tim Horton and Wenson Hsieh.

Currently, when searching a conversation in Mail, find overlays for multiple
messages do not animate in at the same time. This issue occurs due to the fact
that each message is in its own web view, backed by its own web process, and
showing the find overlay involves communication across UI and web processes.

In order to ensure the overlays are animated in/out at the same time, give
Mail access to the find overlay layer. The overlays can then be displayed
alongside other Mail content. The overlay layer is exposed through WKWebView
SPI, and _WKFindDelegate methods to inform the client when the overlay is
added/removed.

Alternatives considered:

1. Mail owns the overlay layer and paints highlights themselves. This approach
was not chosen, as WebKit would still need to provide SPI to vend the rects, and
the complexity of implementation increases as the highlight rects could change
position relative to the web view (for example, in subscrollable regions).
In general, this approach would involve significant work in both Mail and WebKit.

2. Mail owns the overlay layer and gives it to WebKit. This approach was not
chosen, as it is incompatible with the existing remote layer tree.

* Source/WebCore/page/PageOverlay.cpp:
(WebCore::PageOverlay::create):
(WebCore::PageOverlay::PageOverlay):
* Source/WebCore/page/PageOverlay.h:
* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::installPageOverlay):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::supportsLayerType):
* Source/WebCore/platform/graphics/GraphicsLayer.h:

Introduce a new layer type that ensures the layer is always tile-backed. This
is necessary for client-added overlays, since the layer type could otherwise
change, leaving the client with the incorrect layer.

(WebCore::GraphicsLayer::setOpacity):

Export this method, so that the find overlay can begin at zero opacity. The
client is responsible for changing the opacity.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayer::supportsLayerType):
(WebCore::GraphicsLayerCA::initialize):
(WebCore::GraphicsLayerCA::adjustCoverageRect const):
(WebCore::GraphicsLayerCA::requiresTiledLayer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::isTiledBackingLayer const):
* Source/WebKit/UIProcess/API/APIFindClient.h:
(API::FindClient::didAddLayerForFindOverlay):
(API::FindClient::didRemoveLayerForFindOverlay):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFindDelegate.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _processWillSwapOrDidExit]):
(-[WKWebView _didCommitLayerTree:]):

Inform the client that the find overlay is available, once it has been created
and parented. Note that the layer properties arrive in a separate transaction,
since the overlay container is created lazily. Consequently, we wait until the
layer is parented. This ensures that the opacity of the layer is actually zero
once the client is notified.

(-[WKWebView _addLayerForFindOverlay]):

Mark the find overlay as pending once we have a layer ID. At this point, the
layer is not yet in the tree.

(-[WKWebView _removeLayerForFindOverlay]):

Do not send a message to the web process if there is no overlay.

(-[WKWebView _layerForFindOverlay]):

Return the find overlay layer using the `_committedFindLayerID`.

* Source/WebKit/UIProcess/Cocoa/FindClient.h:
* Source/WebKit/UIProcess/Cocoa/FindClient.mm:
(WebKit::FindClient::setDelegate):
(WebKit::FindClient::didAddLayerForFindOverlay):
(WebKit::FindClient::didRemoveLayerForFindOverlay):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addLayerForFindOverlay):
(WebKit::WebPageProxy::removeLayerForFindOverlay):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::addLayerForFindOverlay):

Ensure the overlay is always tile-backed. Without this change, the layer ID can
change depending on the size of the web view, and could leave the client with
the incorrect layer.

(WebKit::WebFoundTextRangeController::removeLayerForFindOverlay):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addFindOverlayLayer):
(WebKit::WebPage::removeFindOverlayLayer):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(-[TestFindDelegate setDidAddLayerForFindOverlayHandler:]):
(-[TestFindDelegate didAddLayerForFindOverlayHandler]):
(-[TestFindDelegate setDidRemoveLayerForFindOverlayHandler:]):
(-[TestFindDelegate didRemoveLayerForFindOverlayHandler]):
(-[TestFindDelegate _webView:didAddLayerForFindOverlay:]):
(-[TestFindDelegate _webViewDidRemoveLayerForFindOverlay:]):
(TEST):

Added a test to exercise the new SPI.

Canonical link: https://commits.webkit.org/252181@main
